core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -539,6 +541,9 @@ class SparkContext(config: SparkConf) extends Logging {
     _heartbeatReceiver = env.rpcEnv.setupEndpoint(
       HeartbeatReceiver.ENDPOINT_NAME, new HeartbeatReceiver(this))
 
+    // Initialize any plugins before the task scheduler is initialized.
+    _plugins = PluginContainer(this)
```

Comment:
There are obvious advantages to initialize the driver plugin at this early stage, however this is not an ideal point for registering metrics (for those plugins that want to do so), as the metrics source should ideally be registered with _env.metricsSystem which is only started at later point, after the task scheduler has been started. As it is now, driver plugin metrics do not get the application id, so they are difficult to consume. 
How about, for example, splitting the driver plugin code, so that the metrics registration part (if any is needed) can take advantage of registering with env.metricsSystem.registerSource(...)?

Reply:
I can take another look, but I'm 99% sure that when I tried `registerSource` after the metrics system had already been initialized, the metrics were not being picked up by `ConsoleSink`.

Maybe it's a bug in that particular sink, but I didn't investigate that far.

Reply:
Also, if you really want to, you can register metrics later. Just install a listener and wait for the "application start" event. (You just need at least a dummy metric registered here, or your source won't be initialized. But that can be fixed easily.)

Reply:
I'll just move the conversation back here because github doesn't have top-level threads and that's annoying.

> Just to clarify, I was thinking about registering the plugin source in the driver somewhere "near" to what was done in #23838

That has the problem I ran into - that's done after `metricsSystem.start()` is called, but in my case metrics added after that were not showing up in `ConsoleSink`.

And the thing is, I can't just move the metrics registration to that spot. I'd have to move all plugin initialization to that spot, otherwise there's no place where the plugin can initialize the metrics before the registration. And I can't move the plugin initialization to that spot, because then I can't send the plugin data to executors via config. I could use RPCs for that but it would slow down executor startup unnecessarily.

So, ignoring the `ConsoleSink` issue, I still think my suggestion is better: if you need to initialize your metrics later, then use a listener that is installed from your plugin's init method.

I can change the code to allow for this late initialization, so you don't have to add a dummy metric in the plugin's init.

Reply:
Ok, here's what happens. If you add a metric to the registry after you register it as a source, those metrics don't show up. At least not in `ConsoleSink`. So it's ok to register the source at the point you mention as long as all metrics have been registered before then.

I'll add a new method to the driver plugins for explicitly registering metrics; that way initialization can happen early, and metrics initialization later. It diverges a bit from the executor API, but I don't see a much better alternative. To be able to register plugin RPC endpoints before executors are up, initialization needs to happen early.

(That also avoids a warning from the metrics system in the output when you register a source before an app ID is known.)

Reply:
I have just tested it and it works for me. Thanks for the work and explanation.